lockin/src/lib.rs

```diff
@@ -0,0 +1,973 @@
+///
+/// `i` - In-phase signal.
+/// `q` - Quadrature signal.
+fn iq_to_a(i: f32, q: f32) -> f32 {
+    2. * sqrt(pow2(i) + pow2(q))
+}
+
+/// Returns angle from in-phase and quadrature signals.
+///
+/// # Arguments
+///
+/// `i` - In-phase signal.
+/// `q` - Quadrature signal.
+fn iq_to_t(i: f32, q: f32) -> f32 {
+    atan2(q, i)
+}
+
+/// Demodulation phase values corresponding to each ADC sample.
+///
+/// # Arguments
+///
+/// * `first_t` - First timestamp value from the current processing
+/// period. The value provided here is ignored if there were no
+/// timestamps in the current processing period.
+/// * `tstamps_mem` - Recorded TimeStamps.
+/// * `phi` - Reference phase offset.
+/// * `fscale` - Frequency scaling factor for the demodulation signal.
+/// * `tadc` - ADC sampling period.
+///
+/// # Latency (ADC batch size = 16): 3.3us
+///
+/// 2.9us from computing `real_phase`.
+fn adc_phases(
+    first_t: u32,
+    tstamps_mem: &mut [TimeStamp; 2],
+    phi: u32,
+    fscale: u32,
+    tadc: u32,
+) -> [f32; SAMPLE_BUFFER_SIZE] {
+    let overflow_count: u32 = tadc * SAMPLE_BUFFER_SIZE as u32;
+    let tref_count: u32 = tstamps_diff(tstamps_mem, overflow_count);
+    let mut thetas: [f32; SAMPLE_BUFFER_SIZE] = [0.; SAMPLE_BUFFER_SIZE];
+    let mut theta_count: u32;
+
+    // 68ns
+    if tstamps_mem[0].sequences_old == 0 {
+        theta_count = tref_count - first_t;
+    } else {
+        theta_count = tstamps_diff(
+            &[
+                TimeStamp {
+                    count: 0,
+                    sequences_old: 0,
+                },
+                tstamps_mem[0],
+            ],
+            overflow_count,
+        );
+    }
+
+    thetas[0] = real_phase(theta_count, fscale, tref_count, phi);
+    for i in 1..SAMPLE_BUFFER_SIZE {
+        theta_count += tadc;
+        thetas[i] = real_phase(theta_count, fscale, tref_count, phi);
+    }
+
+    thetas
+}
+
+/// Number of fast clock counts between two consecutive
+/// TimeStamps. This requires that `tstamps[0]` is more recent than
+/// `tstamps[1]` but otherwise imposes no restrictions on them. For
+/// instance, they can be from different processing periods and this
+/// will still count the number of counts between them, accounting for
+/// overflow wrapping. Both TimeStamps must be valid for this to be
+/// used.
+///
+/// # Arguments
+///
+/// * `tstamps_mem` - Recorded TimeStamps.
+/// * `overflow_count` - Max timestamp value.
+fn tstamps_diff(tstamps_mem: &[TimeStamp; 2], overflow_count: u32) -> u32 {
+    debug_assert!(tstamps_valid_count(tstamps_mem) == 2);
+
+    if tstamps_mem[0].sequences_old == tstamps_mem[1].sequences_old {
+        return tstamps_mem[0].count - tstamps_mem[1].count;
+    }
+
+    let rem0: u32 = tstamps_mem[0].count;
+    let rem1: u32 = overflow_count - tstamps_mem[1].count;
+    let empty_sequences =
+        tstamps_mem[1].sequences_old - tstamps_mem[0].sequences_old - 1;
+
+    rem0 + rem1 + overflow_count * empty_sequences as u32
```

Comment:
Also, if you write that out on paper and simplify it, this becomes clear and easier to read. It also becomes obvious that here you are implementing a number representation and math on it.